### PR TITLE
Adopting the new BaseModel structure in the model files.

### PR DESCRIPTION
**Pull request by davidorme**
# Description

This PR introduces the changes to the actual model files required to adopt the new CoreComponents approach in #374. If you could look at your models to see if all seems sane in the updated files that would be great.

Updated 2024-01-26:

* Tests are updated and  now all pass 
 

### Questions

* Where you are using `layer_roles` in functions, is it worth now using `self.LayerStructure`? That includes things like `n_layers` so it might simplify code to use it directly. Is there anything else to add to it?
* I have aligned all the models to use `self.model_constants`. We now have `self.core_constants` and the models currently use a mix of `model_constants` and `constants`.
* The `BaseModel.__repr__` needed to be updated as `update_interval` is no longer a direct attribute. For some reason this is now displaying update intervals in seconds, but I think we might want a broader rethink of `__repr__` at some point.
* @vgro I have updated the `HydrologyModel.__init__` checking a little - the code wasn't checking that `initial_groundwater_saturation` was numeric.
* @TaranRallings I have updated `AnimalModel.from_config` to create an instance of `AnimalConsts` from the config and then pass it down. It is needed in the main model and at the moment, your functional group definition used a fresh instance that could not be configured.

### Things to do:

- [x] Fix the remaining issues that stop it running.
- [x] Fix the tests.
- [ ] Fix the docs.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] : All tests pass: `$ poetry run pytest`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: virtual_rainforest/models/litter/litter_model.py [hunk 1]
```diff
@@ -146,25 +139,19 @@ def __init__(
 
         self.model_constants = model_constants
```
[the remark this answers is not in the document]
Comment (davidorme) [reply virtual_rainforest/models/litter/litter_model.py:140]: That's my thinking too 😄 We do end up with some very long names, but it is easy to global search and replace long unique names!

File: virtual_rainforest/models/litter/litter_model.py [hunk 2]
```diff
@@ -146,25 +139,19 @@ def __init__(
...
-        self.core_constants = core_constants
-        """Set of constants shared across models."""
 
-        # create a list of layer roles
-        layer_roles = set_layer_roles(canopy_layers, soil_layers)
         # Find first soil layer from the list of layer roles
-        self.top_soil_layer_index = next(
-            i for i, v in enumerate(layer_roles) if v == "soil"
-        )
+        self.top_soil_layer_index: int = self.layer_structure.layer_roles.index("soil")
```
[the remark this answers is not in the document]
Comment (davidorme) [reply virtual_rainforest/models/litter/litter_model.py:144]: I think it was always possible 😄

**Review by vgro — APPROVED** [review]
I see why you made this decision and I agree that the setup, use of layer structure, and model constants are neater now. My only concern is that we maybe overdo it with the wrapping and lose transparency with the core_components. This also means that the unwrapping to use arguments created very long lines which I personally find more difficult to read. I guess that's all a matter of taste, I'm happy for you to merge this change, it's more to keep in mind for further development. Maybe it's time to give the core development another break for the science modules to develop instead of catching up with the structural changes ;-)

**Comment by davidorme** [comment]
> This also means that the unwrapping to use arguments created very long lines which I personally find more difficult to read.

I agree that the lines are long. I think this level of nesting (`BigThing.Container.thing`) is very common in Python but does clash in an ugly way with our long descriptive variable names (which I think are a good thing). One thing that we can do here is unpack `CoreComponents` within an individual model to shorten the reference:

```python
self.update_interval = self.model_timing.update_interval
```

I'd generally argue against doing that for consistency but I don't think it's wrong.

**Comment by davidorme** [comment]
> Maybe it's time to give the core development another break for the science modules to develop instead of catching up with the structural changes ;-)

I'm not sure about this. We've learned a lot about how models work in practice by writing them. I agree that changing the core around makes it harder in the short term but if there are things where we have awkward code in models and have identified a core fix that resolves them, we should fix the core problem, not build on top of the awkward code. 

We definitely shouldn't go mad but we have a core milestone of things that have needed fixing for a while.

**Review by TaranRallings — APPROVED** [review]
The animal changes all look sensible to me!

**Pull request merged**